speculative_decoding/packages/spec_dec.py

```diff
@@ -0,0 +1,369 @@
+            input_ids = np.squeeze(
+                self._tokenizer.encode(input_text, return_tensors="np")
+            )
+        with helpers.timeit(f"Generate({self._model_name}) - input prep", skip=True):
+            inputs = helpers.make_trtllm_inputs(
+                input_ids,
+                max_num_gen_tokens,
+                sampling_config,
+                None,  # draft_tokens
+                self._tokenizer.eos_token_id,
+                self._tokenizer.pad_token_id,
+                self._tokenize_word_list(bad_word_list),
+                self._tokenize_word_list(stop_words_list),
+            )
+        with helpers.timeit(f"Generate({self._model_name})"):
+            result = await self._client.infer(
+                self._model_name, inputs, request_id=request_id
+            )
+            output_ids = helpers.extract_trtllm_outputs(result)
+        with helpers.timeit(f"Generate({self._model_name}) - detokenize", skip=True):
+            output_text = self._tokenizer.decode(output_ids, skip_special_tokens=True)
+        return output_text, output_ids
+
+    async def verify_and_generate(
+        self,
+        confirmed_ids: np.ndarray[int],
+        draft_ids: np.ndarray[int],
+        request_id: str,
+        sampling_config: helpers.SamplingConfig | None = None,
+        bad_word_list: Sequence[str] | None = None,
+        stop_words_list: Sequence[str] | None = None,
+    ) -> tuple[str, np.ndarray[int]]:
+        """Accepts/rejects draft tokens and generates `1` new token.
+
+        If all draft tokens are accepted, the new token extends the the sequence.
+        If some draft tokens are rejected, the new token is generated to replace
+        the first rejected draft token.
+        """
+        num_gen_tokens = len(draft_ids) + 1 if draft_ids is not None else 1
+        # TODO: check len of draft IDs, warn if throw away.
+        with helpers.timeit(f"Generate({self._model_name}) - input prep", skip=True):
+            inputs = helpers.make_trtllm_inputs(
+                confirmed_ids,
+                num_gen_tokens,
+                sampling_config,
+                draft_ids,
+                self._tokenizer.eos_token_id,
+                self._tokenizer.pad_token_id,
+                self._tokenize_word_list(bad_word_list),
+                self._tokenize_word_list(stop_words_list),
+            )
+        with helpers.timeit(f"Verify+Generate({self._model_name})"):
+            result = await self._client.infer(
+                self._model_name, inputs, request_id=request_id
+            )
+            output_ids = helpers.extract_trtllm_outputs(result)
+
+        with helpers.timeit(f"Generate({self._model_name}) - detokenize", skip=True):
+            output_text = self._tokenizer.decode(output_ids, skip_special_tokens=True)
+        return output_text, output_ids
+
+
+QUEUE_SENTINEL = None
+
+
+async def run_speculative_inference(
```

Comment:
Will be really good to have unit tests for this, the code is quite complex. We don't have a standard way of specifying unit tests in a truss right now, but for now we can make assumptions such as use of pytest, writes tests in a top level tests folder and a helper script to run those tests.

Reply:
Added to TODOs at top of file.